src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs

```diff
@@ -31,6 +31,222 @@ public static bool IsAnyNotNullable(this NullableAnnotation annotation)
+        {
+            if (selfAnnotation == otherAnnotation)
+            {
+                return selfAnnotation;
+            }
+
+            if (selfAnnotation.IsAnyNullable() || otherAnnotation.IsAnyNullable())
+            {
+                return selfAnnotation == NullableAnnotation.Nullable || otherAnnotation == NullableAnnotation.Nullable ?
+                            NullableAnnotation.Nullable : NullableAnnotation.NullableBasedOnAnalysis;
+            }
+            else if (selfAnnotation == NullableAnnotation.Unknown)
+            {
+                if (otherAnnotation == NullableAnnotation.Unknown || otherAnnotation == NullableAnnotation.NotNullableBasedOnAnalysis)
+                {
+                    return NullableAnnotation.Unknown;
+                }
+                else
+                {
+                    Debug.Assert(otherAnnotation == NullableAnnotation.NotNullable);
+                    if (isPossiblyNullableReferenceTypeTypeParameter(type))
+                    {
+                        return otherAnnotation;
+                    }
+                    else
+                    {
+                        return NullableAnnotation.Unknown;
+                    }
+                }
+            }
+            else if (otherAnnotation == NullableAnnotation.Unknown)
+            {
+                if (selfAnnotation == NullableAnnotation.NotNullableBasedOnAnalysis)
+                {
+                    return NullableAnnotation.Unknown;
+                }
+                else
+                {
+                    Debug.Assert(selfAnnotation == NullableAnnotation.NotNullable);
+                    if (isPossiblyNullableReferenceTypeTypeParameter(type))
+                    {
+                        return selfAnnotation;
+                    }
+                    else
+                    {
+                        return NullableAnnotation.Unknown;
+                    }
+                }
+            }
+            else
+            {
+                return selfAnnotation == NullableAnnotation.NotNullable || otherAnnotation == NullableAnnotation.NotNullable ?
+                            NullableAnnotation.NotNullable : NullableAnnotation.NotNullableBasedOnAnalysis;
+            }
+        }
+
+        /// <summary>
+        /// Meet two nullable annotations for computing the nullable annotation of a type parameter from upper bounds.
+        /// </summary>
+        public static NullableAnnotation MeetForFixingUpperBounds(this NullableAnnotation a, NullableAnnotation b)
+        {
+            if (a == b)
+            {
+                return a;
+            }
+
+            if (a.IsAnyNullable() && b.IsAnyNullable())
+            {
+                return NullableAnnotation.Nullable;
+            }
+
+            // If nullability on both sides matches - result is that nullability (trivial cases like these are handled before the switch)
```

Comment:
>the switch [](start = 126, length = 10)

Consider adjusting the comment, the switch is no longer part of the implementation.